### ISSUE TITLE
Upgrading to v17 breaks TypeScript types

### ISSUE THREAD

**Issue by willsoto** [octokit/octokit.js#1624]
Hello! 

Here is the error:

```
'Octokit' refers to a value, but is being used as a type here.
```

```ts
import { Octokit } from "@octokit/rest";

class MyClass {             
  constructor(public logger: Logger, public octokit: Octokit) {}
                                                       ^--- error is here
}
```

I think the typings might just be pointing to the wrong thing somehow? I notice in `package.json` that [`types` refers to `index.d.ts`](https://github.com/octokit/rest.js/blob/master/package.json#L65) but when importing the package it is grabbing the types from `node_modules/@octokit/rest/dist-types/index.d.ts`.

When I open that file, this is all it contains:

```ts
import { Octokit as Core } from "@octokit/core";
export declare const Octokit: (new (...args: any[]) => {
    [x: string]: any;
}) & {
    new (...args: any[]): {
        [x: string]: any;
    };
    plugins: any[];
} & typeof Core & import("@octokit/core/dist-types/types").Constructor<void> & import("@octokit/core/dist-types/types").Constructor<{
    paginate: import("@octokit/plugin-paginate-rest").PaginateInterface;
} & import("@octokit/plugin-rest-endpoint-methods/dist-types/generated/types").RestEndpointMethods>;
```

TypeScript: 3.7.5
Node: 12.15
Yarn: 1.22.0

--- 

This code worked prior to v17 and I don't see anything in the changelog regarding changes to types.

Thanks!

**Comment by gr2m** [comment]
Thank you for reporting the issue Will.

I'm not sure if there is anything I did wrong with the TypeScript setup or if it's a problem with the way the package is built with `@pika/build`.

Before, `@octokit/rest` was written in JavaScript and there was an `index.d.ts` TypeScript file. Starting with version 17, everything is written in TypeScript, so I removed the `index.d.ts` files.

Is there something I could/should change in the TypeScript code to make the exported `Octokit` be both the constructor and the type? It'd probably be needed to change in https://github.com/octokit/core.js.

For the time being you should be able to do `typeof Octokit`. I'll add a note to the changelogs

**Comment by gr2m** [comment]
What's odd is that it seems to work for me, I use the `octokit: Octokit` syntax in all my plugins

https://github.com/octokit/plugin-paginate-rest.js/blob/e7c5443dc4126f9826be9d069c5396b55ef98fc9/src/index.ts#L8-L14

**Comment by bcoe** [comment]
The might be some forensic information here for you too: https://github.com/googleapis/repo-automation-bots/pull/311

**Comment by gr2m** [comment]
I _think_ the problem might be the type of `Octokit` as exported by `@octokit/core`

```js
import { Octokit } from `@octokit/core`
```

vs what is returned by `Octokit.plugin()` / `Octokit.defaults()` as it's now happening in the latest `@octokit/rest`:

https://github.com/octokit/rest.js/blob/ead2ce11cae2c86c1da72ce20b9f5b0a9823579b/src/index.ts#L8-L20

The `Octokit` exported by `@octokit/core` seems to be usable as type, while the `Octokit` exported by `@octokit/rest` does not.

Maybe TypeScript doesn't like the `class extends this` in `@octokit/core`?
https://github.com/octokit/core.js/blob/781dc8094528b0a1ff228e29633ed24c997ecacd/src/index.ts#L22-L41

I'd appreciate any help with this. I created a separate package with only the Plugin Architecture bits of Octokit: https://github.com/gr2m/javascript-plugin-architecture-with-typescript-definitions/, it might be easier to create a reproducible test case using that, and ask for more help once we do

**Comment by willsoto** [comment]
Thanks for digging into it! I might be able to spend some time next week looking into it more. I'll report back with any findings

**Comment by jurijzahn8019** [comment]
Hi, also this does not work anymore:

```ts
  public async listTeams(org: Org): Promise<Octokit.TeamsListResponseItem[]> {
    const teams = (await this.client.paginate(
      this.client.teams.list.endpoint.merge({ org: org.org, per_page: 100 }),
      GithubApi.paginationWithLimits
    )) as Octokit.TeamsListResponseItem[];

    return teams;
  }
```

Neither `Octokit.TeamsListResponseItem` nor `TeamsListResponseItem` are exported anymore.

**Comment by gr2m** [comment]
Hmm I think the right place to import endpoint response types from would be https://github.com/octokit/types.ts. It would require some work though. They currently exist in https://github.com/octokit/plugin-rest-endpoint-methods.js/blob/80132aaaa03a61d4b7645cc25a860ad6b02fd427/src/generated/types.ts#L10-L20928 but are not being exported.

For the time being you should be able to infer the type like this:

```ts
type Unwrap<T> = T extends Promise<infer U> ? U : T;

type TeamsListResponseData = Unwrap<
  ReturnType<typeof this.client.teams.list>
>["data"];
```

**Comment by jurijzahn8019** [comment]
This is awesome in terms of "power of typescript" 🤩,
Thank you for the hint! So i did this: 

```ts
export type UnwrapPromise<T> = T extends Promise<infer U> ? U : T;

export type UnwrapArray<T> = T extends Array<infer U> ? U : T;
/**
 * Type of Octokit used instead of Octokit from core
 */
export type Github = InstanceType<typeof Octokit>;
/**
 * This is used to extrapolate Default function signature
 */
export type GithubExampleFunc = Github["teams"]["getByName"];
/**
 * Extrapolated OctokitResponse by remapping Data Payload Type
 * @template T Type of Response payload
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type GithubExampleResponse<T = any> = Omit<
  UnwrapPromise<ReturnType<GithubExampleFunc>>,
  "data"
> & { data: T };
/**
 * Describes default Octokit function signature by mapping
 * endpoint from Example Function
 */
export type GithubFunction = {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  (params?: any): Promise<{ data: any }>;
  endpoint: GithubExampleFunc["endpoint"];
};
/**
 * Extrapolates Octokit default function input extrapolated
 * from Function Signature. Used in Generics and to build
 * Input types for Endpoint Methods
 * @template F type of function to extrapolate params from
 */
export type GithubParams<F extends GithubFunction> = Required<Parameters<F>>[0];

/**
 * Extrapolates datapayload type from a given function return
 */
export type GithubData<F extends GithubFunction> = UnwrapPromise<
  ReturnType<F>
>["data"];

/**
 * Extrapolates a singe item type from a given function return
 */
export type GithubItem<F extends GithubFunction> = UnwrapArray<GithubData<F>>;

/**
 * Extrapolates Octokit default function result extrapolated
 * from Function Signature. Used in Generics and to build
 * Return types for Endpoint Methods
 * @template F type of function to extrapolate params from
 */
export type GithubPromise<F extends GithubFunction> = Promise<GithubData<F>>;
```
and then this: 

```ts
  public paginate<F extends GithubFunction, P extends GithubParams<F>>(
    func: F,
    params: P
  ): GithubPromise<F> {
    return this.client.paginate(
      func.endpoint.merge(params) as Required<
        ReturnType<typeof func.endpoint.merge>
      >,
      GithubApi.paginationWithLimits
    );
  }

  public async listTeams(org: Org): GithubPromise<Github["teams"]["list"]> {
    const teams = await this.paginate(this.client.teams.list, {
      org: org.org,
      per_page: 100
    });
    return teams;
  }
```

But is hope we can use exported types soon 🙏 😀

**Comment by gr2m** [comment]
@jurijzahn8019 thank you for sharing! I wonder if we should export TypeScript helper functions like that to make it easy for people to extract response types from methods such as `octokit.teams.list` or routes such as `GET /orgs/:org/teams` 🤔 

What would some "dream code" look like from your perspective? Maybe something like this?

```ts
import type { ResponseTypes, PaginationItemTypes } from "@octokit/types";
```

Then you could use

```ts
ResponseTypes["teams"]["list"]
```

for the entire response, or

```ts
PaginationItemTypes["teams"]["list"]
```

to get the type of a single item as it would be returned by the paginate method?

**Comment by ianwremmel** [comment]
Maybe this is a naive question, but if the various methods can return these types, what makes it so difficult to export them? Why does a builtin solution require type-gymnastics to extract them?

**Comment by gr2m** [comment]
> what makes it so difficult to export them

it's not difficult, it just has to be done. Octokit v17 was a major refactoring. It's ~10LOC now:
https://github.com/octokit/rest.js/blob/master/src/index.ts

There are no type definitions in this repository any longer, it just combines `@octokit/core` and plugins into a REST API best practice SDK. The endpoint methods are added via https://github.com/octokit/plugin-rest-endpoint-methods.js/. I still think extracting the types from methods is the cleanest way, because there is no chance that the exported types and what the methods return become out of sync

**Comment by gr2m** [comment]
Maybe I'll do sth like this

```ts
import type { MethodToResponseType } from "@octokit/types";

type TeamsListResponse = MethodToResponseType<typeof octokit.teams.list>
```

**Comment by jurijzahn8019** [comment]
Although this works, the usage is very cryptic. The IDE will not help you. I am using the types I created, but they are not a replacement for real typings.

**Comment by disintegrator** [comment]
It would be nice if `@octokit/rest` exported a proper type for the client. Until then you can work around this issue by doing the following:

```ts
import { Octokit } from "@octokit/rest";

type OctokitClient = InstanceType<typeof Octokit>
```

**Comment by gr2m** [comment]
@disintegrator do you have an idea why that is? I'd appreciate help to fix / workaround that problem

**Comment by disintegrator** [comment]
@gr2m in the release notes you mention to users that they should switch to `typeof Octokit` since the library no longer exports a type. `typeof Octokit` returns a crazy _constructor_ type due to the new plugin based rewrite. Anyways, my guess is that people don't care much for the constructor. They want to be able to use the instance type when declaring functions that accept the octokit client for example. This is why I'd advise them to use `InstanceType<typeof Octokit>`.

Alternatively, I can create a PR that exports the following in `src/index.ts`:

```diff
import { Octokit as Core } from "@octokit/core";
import { requestLog } from "@octokit/plugin-request-log";
import { paginateRest } from "@octokit/plugin-paginate-rest";
import { restEndpointMethods } from "@octokit/plugin-rest-endpoint-methods";

import { VERSION } from "./version";

export const Octokit = Core.plugin([
  (requestLog as unknown) as () => void,
  restEndpointMethods,
  paginateRest
]).defaults({
  userAgent: `octokit-rest.js/${VERSION}`
});

+export type OctokitRestInstance = InstanceType<typeof Octokit>;
```

**Comment by jamescdavis** [comment]
@disintegrator how about giving the exported type the same name (`Octokit`), so you don't have to do a separate import for it:
```diff
import { Octokit as Core } from "@octokit/core";
import { requestLog } from "@octokit/plugin-request-log";
import { paginateRest } from "@octokit/plugin-paginate-rest";
import { restEndpointMethods } from "@octokit/plugin-rest-endpoint-methods";

import { VERSION } from "./version";

export const Octokit = Core.plugin([
  (requestLog as unknown) as () => void,
  restEndpointMethods,
  paginateRest
]).defaults({
  userAgent: `octokit-rest.js/${VERSION}`
});

+export type Octokit = InstanceType<typeof Octokit>;
```

**Comment by gr2m** [comment]
> you mention to users that they should switch to typeof Octokit since the library no longer exports a type

I added it later, this was not intended. I think this is caused by a TypeScript bug.

> how about giving the exported type the same name

@jamescdavis  oh that works? That'd be great! Would love a PR! Could you maybe add a failing test first to show the current state, then push the fix in a separate commit once the tests were triggered on the CI?

**Comment by jamescdavis** [comment]
Yep, that works because you can export a value and a type with the same name and TS will use the correct one based on how it's used.

I opened a PR before I saw your comment, but yeah, a test would be a good idea. I'll back out the fix, add a test, let it fail, then add the fix again.

**Comment by bcoe** [comment]
I sat down again to try to spend a bit of time upgrading libraries that use `octokit/rest.js`, and am a bit at a loss. Historically I'd been using this approach:

```typescript
    const pullsResponse = (await this.request(
      `GET /repos/:owner/:repo/pulls?state=open&per_page=${perPage}${
        this.proxyKey ? `&key=${this.proxyKey}` : ''
      }`,
      {
        owner: this.owner,
        repo: this.repo,
      }
    )) as Octokit.Response<Octokit.PullsListResponseItem[]>;
```

To perform a request to an arbitrary GitHub endpoint, and provide a type hint about the response that I expect back.

With `v17.x.x` I can't find similar type hints exported anywhere.

**Comment by gr2m** [comment]
@bcoe this is unrleated to the original issue.

I will do two things that will help

1. Add response types to https://github.com/octokit/types.ts/
2. Add types to `octokit.request`, just like they already exist for `octokit.endpoint`.

By the way you can simplify your code

```js
const pullsResponse = (await this.request(
  'GET /repos/:owner/:repo/pulls',
  {
    owner: this.owner,
    repo: this.repo,
    state: 'open'
    per_page: perPage,
    key: this.proxyKey
  }
)) as Octokit.Response<Octokit.PullsListResponseItem[]>;
```

What is the `key` parameter?

**Comment by bcoe** [comment]
@gr2m we have an internal proxy that takes a special key for requests, which was originally why I was using `this.request` directly, rather than `octokit.pulls.list`.

Hoping that I might be able to get on to `v17` without a full rewrite 😄 

I appreciate you adding these types to `@octokit/types` 👌

**Comment by prokopsimek** [comment]
The same issue here in https://github.com/DXHeroes/dx-scanner/pull/266.

I am unable to set:
```typescript
const params: Octokit.PullsListParams = { owner, repo };
// and
const params: Octokit.IssuesListForRepoParams = { owner, repo };
// and
const params: Octokit.IssuesListCommentsParams = { owner, repo, issue_number: issueNumber };
// and
const params: Octokit.PullsListCommitsParams = { owner, repo, pull_number: prNumber };
// and
private unwrap<T>(clientPromise: Promise<Octokit.Response<T>>): Promise<Octokit.Response<T>> { ... }
```

Could you help me how can I change the code according to `v17.1` changes?

Thanks

**Comment by jurijzahn8019** [comment]
Hi @prokopsimek, 

this could work

```ts
import type {
  PullsListParams,
  IssuesListForRepoParams,
  IssuesListCommentsParams,
  PullsListCommitsParams,
} from "@octokit/plugin-rest-endpoint-methods/dist-types/generated/types";
import type { OctokitResponse } from "@octokit/types";

const owner = "foo";
const repo = "foo";
const issueNumber = 1;
const prNumber = 2;

const params1: PullsListParams = { owner, repo };
// and
const params2: IssuesListForRepoParams = { owner, repo };
// and
const params3: IssuesListCommentsParams = {
  owner,
  repo,
  issue_number: issueNumber,
};
// and
const params4: PullsListCommitsParams = {
  owner,
  repo,
  pull_number: prNumber,
};
// and

export class Foo {
  private unwrap<T>(
    clientPromise: Promise<OctokitResponse<T>>
  ): Promise<OctokitResponse<T>> {}
}

```

**Comment by jurijzahn8019** [comment]
HI @gr2m, please correct me, but i cannot import any Response Type for example: `Octokit.ReposGetContentsResponse` or `Octokit.ActionsGetPublicKeyResponse`

Are they exported somewhere else?

**Comment by gr2m** [comment]
not yet, follow this issue: https://github.com/octokit/types.ts/issues/25

**Comment by prokopsimek** [comment]
@jurijzahn8019 Thanks! It works. 🎉

**Comment by gr2m** [comment]
note that if you import from `"@octokit/plugin-rest-endpoint-methods/dist-types/generated/types"`, you use internal APIs that might break at any point. I'd advise to pin the version of `@octokit/plugin-rest-endpoint-methods` in your package.json, or extract the types from the `octokit.*` endpoint methods as mentioned above in https://github.com/octokit/rest.js/issues/1624#issuecomment-590436498

**Comment by gr2m** [comment]
I've an open pull request for `@octokit/types` that will add two helpers

1. GetResponseType
2. GetResponseDataType

https://github.com/octokit/types.ts/pull/32

These allow to extract the response type / response data type from the endpoint methods.

Example:

```ts
import { GetResponseTypeFromEndpointMethod, GetResponseDataTypeFromEndpointMethod } from "@octokit/types";
import { Octokit } from "@octokit/rest";

const octokit = new Octokit();
type CreateLabelResponseType = GetResponseTypeFromEndpointMethod<
  typeof octokit.issues.createLabel
>;
type CreateLabelResponseDataType = GetResponseDataTypeFromEndpointMethod<
  typeof octokit.issues.createLabel
>;
```

I will create a separate pull request which will export a type for each response type in a separate PR. I'm not yet sure if I'll export the response type in `@octokit/types` or `@octokit/plugin-rest-endpoint-methods`, or both

Feedback welcome 😁

**Comment by jurijzahn8019** [comment]
The best way in my opinion, would be to re-export the types via @octokit/rest. 

Otherwise people will need to install other packages explicitly.

**Comment by gr2m** [comment]
`@octokit/types` is already in the dependency tree, you shouldn't install it separately.

The only use case I've found when people want to access the response types is for `octokit.paginate`, and I'll try to fix that with the method itself. If anyone can think of other use cases please let me know.

**Comment by jurijzahn8019** [comment]
Technically you're right. But there are two inconveniences with this aproach:

1. IDE does not help you to discover the deps tree:

![octokit_imports](https://user-images.githubusercontent.com/44019949/78748475-6e2b9400-796c-11ea-9eb3-600afad9cd90.png)

2. Eslint complains

![octokit_eslint](https://user-images.githubusercontent.com/44019949/78748504-7b488300-796c-11ea-8df6-1db8d06cd8ed.png)

I Understand, this is not something what will cause the end of the world, but it disturbs developer experience. And i think developers are your main customer group. So why not to make things easier :smile:

**Comment by gr2m** [comment]
One reason to not export the response types is that it adds a huge API surface that I must maintain and prevent accidental breaking changes in. And the 2nd reason is that I still don't understand why it's necessary in the first place, besides the mentioned use case with `octokit.paginate`, that I'll resolve properly

**Comment by jurijzahn8019** [comment]
Something like

```ts
export * from "@octokit/types"
```

will not work?

**Comment by gr2m** [comment]
no, every breaking change in `@octokit/types` would cause a breaking change in `@octokit/rest`

**Comment by jurijzahn8019** [comment]
Ok, i do not see how this will propagate, if people do not use reexported types.
But as long i can install `@octokit/types` / `@octokit/plugin-rest-endpoint-methods` to get what i need, i will be ok. Go for it :+1:

**Comment by gr2m** [comment]
Do you use the types for anything other than pagination?

Instead of 

```ts
const teams = (await octokit.paginate(
  octokit.teams.list.endpoint.merge({ org: org.org, per_page: 100 })
)) as Octokit.TeamsListResponseItem[];
```

I'll probably implement something like this


```ts
const teams = await octokit.paginate(
  octokit.teams.list,
  { org: org.org, per_page: 100 }
)
```

With the types for `teams` being set correctly

**Comment by jurijzahn8019** [comment]
I have a wrapper with exports methods like:

```ts
  public async listReposByTeam(
    orgTeam: OrgTeam
  ): GithubPromise<Github["teams"]["listReposInOrg"]> {
    ...
  }
```

and then in many places i process results like:

```ts
public async processRepo(
    repo: GithubItem<Octokit["teams"]["listReposInOrg"]>
  ): Promise<...> {
...
}
```

**Comment by ianwremmel** [comment]
> Do you use the types for anything other than pagination?

One example from my codebase: I have common code that may be called as the result of receiving a `marketplace_purchase` webhook or as a result of requesting subscriptions from the api. That code accepts are parameter typed as `WebhookPayloadMarketplacePurchase` | `AppsListAccountsUserOrOrgOnPlanResponseItem`. 

As another example, I have a chunk of code that prepares `ChecksUpdateResponse` but hands it off to a background process for actually submitting it to the API, so the code that does the final submission needs to specify `ChecksUpdateResponse` for one of its function parameters` types.

I've been blocked on upgrading from v16 because I explicitly use Octokit's types extensively in my codebase.

I do understand your point about not creating an always-breaking-change dependency between `@octokit/types` and `@octokit/rest`, but this certainly makes the project much harder to use for my use case.

I wonder if there's some typescript magic like `GetResponseTypeFromEndpointMethod` that could produce the full set of types with just one invocation...

**Comment by ianwremmel** [comment]
This helps a little bit to simplify extracting types (might be the wrapper that @jurijzahn8019 was talking about), but it makes it impossible to (e.g. in vscode) use F12 to get to the original definition of the type to look at its properties.

```ts
import {Octokit} from '@octokit/rest';
type OctokitInstanceType = InstanceType<typeof Octokit>;
type GetResponseTypeFromOctokitMethodPath<
  O extends OctokitInstanceType,
  P extends keyof O,
  M extends keyof O[P]
> = GetResponseDataTypeFromEndpointMethod<O[P][M]>;
```

**Comment by gr2m** [comment]
I have the response types in an internal, generated file at https://raw.githubusercontent.com/octokit/plugin-rest-endpoint-methods.js/master/src/generated/types

They look like this:

```ts
export type RestEndpointMethods = {
  actions: {
    createRegistrationToken: {
      (
        params?: RequestParameters & ActionsCreateRegistrationTokenParams
      ): Promise<OctokitResponse<ActionsCreateRegistrationTokenResponse>>;

      endpoint: EndpointInterface;
    };
  }
}
```

Say we would want to export something like

```
import { EndpointMethodResponseTypes } from "@octokit/plugin-rest-endpoint-methods.js"
```

Which would be a map so you could access response types like this

```ts
EndpointMethodResponseTypes["actions"]["createRegistrationToken"]
```

We could re-export the `EndpointMethodResponseTypes` from `@octokit/rest` for convenience.

Would that be helpful? 

How would we generate the `EndpointMethodResponseTypes` type similar to @ianwremmel's [example above](https://github.com/octokit/rest.js/issues/1624#issuecomment-611300047)?

**Comment by jurijzahn8019** [comment]
What we are really missing is 


```ts
import {​
ActionsCreateRegistrationTokenParam​,
ActionsCreateRegistrationTokenResponse
 } from "@octokit/whatever"

```

**Comment by ianwremmel** [comment]
yep, exactly. I think the helper I posted is a very slight improvement over the existing type helpers, but it's still pretty tedious and, as I mentioned, doesn't integrate well with IDEs. I think we really do need a named importable for each set of parameters and response.

> no, every breaking change in `@octokit/types` would cause a breaking change in `@octokit/rest`

Isn't that effectively true anyway? If the rest API changes in a way that would cause `@octokit/types` to have a breaking change, shouldn't that also imply that a method's parameters or result in `@octokit/rest` has also experienced a breaking change?

**Comment by gr2m** [comment]
> If the rest API changes

The REST API itself does not change. What does change much more often are the titles in the documentation and hence the type names which are derived from that, just like the method names. I do have deprecation for the method names in place, I'm not sure if I want to do the same for all the types, too. It's probably going to be thousands of additional exports if I exported all the generated types. I fear it will increase the maintenance burden a lot, and I already have very little time left to do anything other than maintenance on Octokit.

> ActionsCreateRegistrationTokenParam​

I understand the need for response types. But what do you need the parameter types for?

**Comment by ianwremmel** [comment]
I've wrapped calls to octokit methods. One such example:

```ts
  async updateCheck(
    repoSelector: ClientSelector,
    {
      label: name,
      ...params
    }: Omit<Octokit.ChecksUpdateParams, 'name'> & {label?: string}
  ): Promise<Octokit.ChecksUpdateResponse> {
```

Arguably, that should be 

```ts
  async updateCheck(
    repoSelector: ClientSelector,
    payload: Octokit.ChecksUpdateParams
  ): Promise<Octokit.ChecksUpdateResponse> {
```

but that doesn't really change the need.

I've written a service class that manages octokit instances so the rest of the codebase doesn't need to deal with the vagaries of application-specific credentials, installation-specific credentials, and user-specific credentials. The only way to call the github api in this code base is to call a method on that service which knows how to retrieve or instantiate the right octokit instance. Subsequently, many of those methods need parameter types from Octokit.

**Comment by gr2m** [comment]
Could you make a separate issue describing your use case with more code examples? I'd like to understand why/if wrapping the Octokit methods is really necessary

**Comment by robsonmeemo** [comment]
I need to create a list of ChecksCreateParamsOutputAnnotations that I'm going to eventually pass to checks.create(), but I can't store them in the proper type. I don't understand why types that are part of the API are not exported. Is there a workaround that I'm missing?

**Comment by gr2m** [comment]
@robsonmeemo please follow discussion at https://github.com/octokit/rest.js/discussions/1697

**Comment by robsonmeemo** [comment]
I'm using the solution I found here: https://github.com/octokit/rest.js/issues/1668#issuecomment-620970514

**Comment by peter-evans** [comment]
Admittedly I'm fairly new to Typescript, but I'm completely baffled how to get the `PullsListParams` type in v17. Tried to use the method outlined [here](https://github.com/octokit/rest.js/issues/1624#issuecomment-602689919), but it didn't work, perhaps because the API has changed. Tried to wrap my head around this [kind of solution](https://github.com/octokit/rest.js/issues/1624#issuecomment-590436498), but couldn't figure it out.

```
    const [owner, repo] = repository.split('/')
    const params: Octokit.PullsListParams = {
      owner: owner,
      repo: repo,
      state: 'open'
    }
    if (head.length > 0) params.head = head
    if (base.length > 0) params.base = base
    const {data: pulls} = await this.octokit.pulls.list(params)
```

**Update:** I switched to use [@octokit/graphql](https://github.com/octokit/graphql.js) instead and found the parameter type I needed for that like this.
```
import * as OctokitTypes from '@octokit/types'

const params: OctokitTypes.RequestParameters = {
  owner: owner,
  repo: repo
}
```

**Comment by JackFreelander** [comment]
> no, every breaking change in `@octokit/types` would cause a breaking change in `@octokit/rest`

But this is the nature of a type'd language and is handled with proper versioning.  I am trying to upgrade from @octokit/rest 16.36.0 -> 18.0.0 and it's a total mess.  I've got hundreds of TypeScript errors now about conflicts between types and namespaces, and I use the Octokit types (which seem to be buried in @octokit/types) like `Octokit.PullsGetResponse` and `Octokit.ReposListCommitsResponseItem[]` and countless others, and none of them seem to be properly exposed here.

Can you provide sample code for how consumers are expected to import the namespace and its types?
